src/io/linux.zig

```diff
@@ -215,9 +300,128 @@ pub const IO = struct {
         };
         completion.prep(sqe);
 
+        if (self.awaiting) |first| {
+            assert(self.awaiting_count > 0);
+            assert(first.awaiting_previous == null);
+            first.awaiting_previous = completion;
+            completion.awaiting_next = first;
+        } else {
+            assert(self.awaiting_count == 0);
+        }
+        self.awaiting = completion;
+        self.awaiting_count += 1;
+
         self.ios_queued += 1;
     }
 
+    /// Cancel should be invoked at most once, before any of the memory owned by read/recv buffers
+    /// is freed (so that lingering async operations do not write to them).
+    ///
+    /// After this function is invoked:
+    /// - No more completion callbacks will be called.
+    /// - No more IO may be submitted.
+    ///
+    /// This function doesn't return until either:
+    /// - All events submitted to io_uring have completed.
+    ///   (They may complete with `error.Canceled`).
+    /// - Or, an io_uring error occurs.
+    ///
+    /// TODO(Linux):
+    /// - Linux kernel ≥5.19 supports the IORING_ASYNC_CANCEL_ALL and IORING_ASYNC_CANCEL_ANY flags,
+    ///   which would allow all events to be cancelled simultaneously with a single "cancel"
+    ///   operation, without IO needing to maintain the `awaiting` doubly-linked list and the `next`
+    ///   cancellation stage.
+    /// - Linux kernel ≥6.0 supports `io_uring_register_sync_cancel` which would remove the `queued`
+    ///   cancellation stage.
+    pub fn cancel(self: *IO) !void {
+        assert(self.cancel_status == .inactive);
+        self.awaiting_verify();
+
+        // Ensure that even if we return early due to an io_uring that IO won't accept more IO.
+        defer self.cancel_status = .done;
+
+        self.cancel_status = .next;
+
+        // Discard any operations that haven't started yet.
+        while (self.unqueued.pop()) |_| {}
```

Comment:
Ahh, I see it now, we actually arrange it so that the callback is _not_ called on cancellation. That is, if we call cancel, and at that time there's _already_ a completion in the completion queue (i.e., the operation actually fully finished before cancelation), the callback for that is not called. 

This seems reasonable, I think, to make sure that individual callbacks won't have to worry about the cancelation! And, if the callback holds some closure on the client side, I assume it's up to the client to properly clean it up!

Reply:
Exactly! Once cancellation starts, we don't call any IO callbacks. This reduces the surface area added by cancellation -- message_bus/storage don't really need to concern themselves with it. This "all-or-nothing" approach to cancellation simplifies the mental model, no need to worry about callbacks starting up additional IO work.